ophyd/flyers.py

```diff
@@ -105,10 +166,151 @@ def collect(self):
+
+
+class MonitorFlyerMixin(BlueskyInterface):
+    '''A bluesky-compatible flyer mixin, using monitor_attrs
+
+    At kickoff(), all monitor_attrs will be subscribed to and monitored for the
+    until complete() is called. `complete` returns a DeviceStatus instance,
+    which indicates when the data is ready to be collected.  The acquired
+    values are then be retrievable as bluesky bulk-readable documents in
+    collect().
+
+    Parameters
+    ----------
+    monitor_attrs : list, optional
+        List of signal attribute names to monitor
+    '''
+    def __init__(self, *args, monitor_attrs=None, stream_name=None, **kwargs):
+        if monitor_attrs is None:
+            monitor_attrs = []
+        self.monitor_attrs = monitor_attrs
+        self.stream_name = stream_name
+        self._acquiring = False
+        self._paused = False
+        self._collected_data = None
+
+        super().__init__(*args, **kwargs)
+
+    def kickoff(self):
+        '''Start collection
+
+        Returns
+        -------
+        DeviceStatus
+            This will be set to done when acquisition has begun
+        '''
+        self._collected_data = OrderedDefaultDict(lambda: {'values': [],
+                                                           'timestamps': []})
+        self._start_time = ttime.time()
+        self._acquiring = True
+        self._paused = False
+        self._add_monitors()
+        st = DeviceStatus(self)
+        st._finished(success=True)
+        return st
+
+    def _add_monitors(self):
+        for attr in self.monitor_attrs:
+            obj = getattr(self, attr)
+            if isinstance(obj, Device):
+                raise ValueError('Cannot monitor Devices, only Signals.')
+            obj.subscribe(functools.partial(self._monitor_callback,
+                                            attribute=attr))
+
+    def _monitor_callback(self, attribute=None, obj=None, value=None,
+                          timestamp=None, **kwargs):
+        '''A monitor_attr signal has changed'''
+        if not self._acquiring or self._paused:
+            return
+
+        if value is None or timestamp is None:
+            data = obj.read()[obj.name]
+            value = data['value']
+            timestamp = data['timestamp']
+
+        collected = self._collected_data[attribute]
+        collected['values'].append(value)
+        collected['timestamps'].append(timestamp)
+
+    def describe_collect(self):
+        '''Description of monitored attributes retrieved by collect'''
+        # single stream?
+        desc = OrderedDict()
+        for attr in self.monitor_attrs:
+            desc.update(self._describe_attr_list([attr]))
+        return {self.stream_name: desc}
+
+    def _clear_monitors(self):
+        '''Clear all subscriptions'''
+        for attr in self._collected_data.keys():
+            obj = getattr(self, attr)
+            try:
+                obj.clear_sub(self._monitor_callback)
```

Comment:
I do not think this will actually remove the callback.  Above you add a partial of the callback, this tries to remove the method (unless partial is smarter than I think).